core/storage/question/gae_models.py

```diff
@@ -260,6 +260,85 @@ def get_question_skill_links_by_skill_ids(
         )
         return question_skill_link_models, next_cursor_str
 
+    @classmethod
+    def get_question_skill_links_based_on_mastery_equidistributed_by_skill(
+            cls, total_question_count, skill_ids, degrees_of_mastery):
+        """Fetches the list of constant number of QuestionSkillLinkModels
+        linked to the skills, sorted by the difference between skill
+        difficulty and skill mastery.
+
+        Args:
+            total_question_count: int. The number of questions expected.
+            skill_ids: list(str). The ids of skills for which the linked
+                question ids are to be retrieved.
+            degrees_of_mastery: dict(str, float). The keys are the requested
+                skill IDs. The values are the corresponding mastery degree of
+                the user.
+
+        Returns:
+            list(QuestionSkillLinkModel). A list of QuestionSkillLinkModels
+                corresponding to given skill_ids, with
+                total_question_count/len(skill_ids) number of questions for
+                each skill. If not evenly divisible, it will be rounded up.
+                If not enough questions for a skill, just return all questions
+                it links to. The order of questions will follow the order of
+                given skill ids, and the order of questions for the same skill
+                follows the difference between skill difficulty and mastery.
+        """
+        question_count_per_skill = int(
+            math.ceil(float(total_question_count) / float(len(skill_ids))))
+
+        question_skill_link_models = []
+        existing_question_ids = []
+
+        for skill_id in skill_ids:
+            degree_of_mastery = degrees_of_mastery[skill_id]
```

Comment:
This will throw an exception is degrees_of_mastery is a None value or if skill_id doesn't exist in the map (See the following example: http://www.codeskulptor.org/#user46_EjJOe9mKgs_1.py)

Reply:
Done.

Reply:
Wait I think you didn't get me, we shouldn't throw an exception if skill_id doesn't exist..there can be cases where the user doesn't have skill mastery for a particular skill_id, what I meant is you need to handle those cases - i.e. set the degree_of_mastery to be 0 if skill_id does not exist in degrees_of_mastery

Reply:
Even though there is no skill mastery for one skill id, the id will still exist in the dict. For example, if "skill_id_1" does not have mastery, then the dict will be
`{skill_id_1: None, skill_id_2: 0.3}`